src/lib/operators/abstract_aggregate_operator.hpp

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include "expression/aggregate_expression.hpp"
+#include "operators/abstract_operator.hpp"
+#include "operators/abstract_read_only_operator.hpp"
+#include "type_comparison.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+/**
+ * Aggregates are defined by the column (ColumnID for Operators, LQPColumnReference in LQP) they operate on and the aggregate
+ * function they use. COUNT() is the exception that doesn't use a column, which is why column is optional
+ * Optionally, an alias can be specified to use as the output name.
+ */
+struct AggregateColumnDefinition final {
+  AggregateColumnDefinition(const std::optional<ColumnID>& column, const AggregateFunction function)
+      : column(column), function(function) {}
+
+  std::optional<ColumnID> column;
+  AggregateFunction function;
+};
+
+/*
+The AggregateFunctionBuilder is used to create the lambda function that will be used by
+the AggregateVisitor. It is a separate class because methods cannot be partially specialized.
+Therefore, we partially specialize the whole class and define the get_aggregate_function anew every time.
+*/
+template <typename ColumnDataType, typename AggregateType, AggregateFunction function>
+struct AggregateFunctionBuilder {
+  void get_aggregate_function() { Fail("Invalid aggregate function"); }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Min> {
+  auto get_aggregate_function() {
+    return [](const ColumnDataType& new_value, std::optional<AggregateType>& current_aggregate) {
+      if (!current_aggregate || value_smaller(new_value, *current_aggregate)) {
+        // New minimum found
+        current_aggregate = new_value;
+      }
+      return *current_aggregate;
+    };
+  }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Max> {
+  auto get_aggregate_function() {
+    return [](const ColumnDataType& new_value, std::optional<AggregateType>& current_aggregate) {
+      if (!current_aggregate || value_greater(new_value, *current_aggregate)) {
+        // New maximum found
+        current_aggregate = new_value;
+      }
+      return *current_aggregate;
+    };
+  }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Sum> {
+  auto get_aggregate_function() {
+    return [](const ColumnDataType& new_value, std::optional<AggregateType>& current_aggregate) {
+      // add new value to sum
+      if (current_aggregate) {
+        *current_aggregate += new_value;
+      } else {
+        current_aggregate = new_value;
+      }
+    };
+  }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Avg> {
```

Comment:
I can't follow. Could you rephrase that?

Reply:
I believe what @Miroka96 wants to say is that, in the end, we need to do the division step to get the average anyway, so there is no need for this average-builder when we already have exactly this functionality in the sum-version.

I believe it would work without this builder, and maybe reduce compile time a little bit, but this might come at the cost of worse code readability. But this is probably a matter of taste. Can we get an official opinion on that? @Bensk1 @mrks ?

Reply:
I think I got it. The current way appears to be more readable; I'd stick with that.